### PR TITLE
fix: ensures that DecCoin doesn't loose precision for very long numbers

### PR DESCRIPTION
**Pull request by stalniy**
## Why

Current implementation relies on [`parseInt` + devision](https://github.com/akash-network/akash-api/blob/main/ts/src/patch/cosmos/base/v1beta1/coin.ts#L29) which will loose precision for very long numbers

## What

1. Rewrites DecCoin encode/decode to relying on BigInt checks + string manipulation
2. Adds more test cases to validate edge cases

### CODE REVIEW

**Review by ygrishajev — DISMISSED** [review]

File: ts/package.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
     ]
   },
   "dependencies": {
+    "@cosmjs/math": "^0.33.0",
```
Comment (troian) [inline ts/package.json:44]: need to check if this version is compatible with sdk-47.
go implementation uses math package of version v1.3.0, which is not latest

File: ts/src/patch/cosmos/base/v1beta1/coin.ts [hunk 2]
```diff
@@ -1,20 +1,18 @@
+import { Decimal } from "@cosmjs/math";
 import * as minimal from "protobufjs/minimal";
 import { Reader } from "protobufjs/minimal";
 
 import * as coin from "../../../../generated/cosmos/base/v1beta1/coin.original";
 import { DecCoin } from "../../../../generated/cosmos/base/v1beta1/coin.original";
 
 const originalEncode = coin.DecCoin.encode;
+const PRECISION = 18;
```
Comment (troian) [inline ts/src/patch/cosmos/base/v1beta1/coin.ts:9]: Lets add a comment the value has been derrived from here https://github.com/cosmos/cosmos-sdk/blob/25b14c3caa2ecdc99840dbb88fdb3a2d8ac02158/math/dec.go#L21

File: ts/package.json (hunk 1 above)
Comment (stalniy) [reply ts/package.json:44]: do you have recommendation how to check it? All test cases from cosmos testdata are passed

**Review by troian — APPROVED** [review]

**Review by troian — CHANGES_REQUESTED** [review]
can you rebase it on to `sdk-47` branch

**Comment by stalniy** [comment]
just to double check, @troian are you asking to change PR target branch to  be sdk-47 instead of main?

**Comment by troian** [comment]
correct, those changes are comming to effect on sdk-47 upgrade

**Comment by stalniy** [comment]
well, this is now impossible because you merged my previous PR which relies on different tool to generate types. And patching should be done in a different way. The aim of this PR was to fix the current sdk version 

Anyway, thanks to @baktun14, I can talk to telescope owners/developers, so potentially they will do the fix on their side and will not need to use patching and I will change the implementation in sdk-47 branch. So, let's wait for some time
